network/rcmgr.go

```diff
@@ -0,0 +1,116 @@
+package network
+
+import (
+	"errors"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var (
+	ErrResourceLimitExceeded = errors.New("resource limit exceeded")
+	ErrResourceScopeClosed   = errors.New("resource scope closed")
+)
+
+// ResourceManager is the interface to the network resource management subsystem
+type ResourceManager interface {
+	// GetProtocolScope retrieves the resource management scope for a specific protocol.
+	// If there is no configured limits for a particular protocol, then the default scope is
+	// returned.
+	GetProtocolScope(protocol.ID) ProtocolScope
+	// GetPeerScope retrieces the resource management scope for a specific peer.
+	GetPeerScope(peer.ID) PeerScope
+
+	// OpenConnection creates a connection scope not yet associated with any peer
+	OpenConnection(dir Direction, usefd bool) (ConnectionScope, error)
+
+	// Close closes the resource manager
+	Close() error
+}
+
+// ResourceScope is the interface for all scopes.
+type ResourceScope interface {
+	// ReserveMemory reserves memory/buffer space in the scope.
+	ReserveMemory(size int) error
+	// ReleaseMemory explicitly releases memory previously reserved with ReserveMemory
+	ReleaseMemory(size int)
+
+	// GetBuffer reserves memory and allocates a buffer through the buffer pool.
+	GetBuffer(size int) ([]byte, error)
+	// GrowBuffer grows a previous allocated buffer, reserving the appropriate memory space.
+	GrowBuffer(buf []byte, newsize, copy int) ([]byte, error)
```

Comment:
What's with the `copy` param?

Reply:
to atomically copy the old buffer while growing/replacing it with a new one;  need to document this.

Reply:
I'm still confused. Can't you just copy the old buffer anyways?

Reply:
Is the idea that stream muxers would use this?
I'm a bit worried about the API here, for example, this doesn't really match the quic-go API.

Reply:
I would like to make it atomically release and give you a new one. It really is a shortcut for Get new, Release old, without oversubscribing memory.

Maybe we dont need it, but of it is an old buffer growing by 1.5x it seems ugood to have it.

Reply:
removed the ncopy parameter, implementation should unconditionally copy.